lib/elixir/lib/module/types/descr.ex

```diff
@@ -495,4 +527,451 @@ defmodule Module.Types.Descr do
+  end
+
+  # Intersects a map with a union of maps.
+  defp map_intersection_aux({tag1, pos1, negs1}, dnf2) do
+    Enum.reduce(dnf2, [], fn {tag2, pos2, negs2}, acc ->
+      try do
+        {tag, fields} = map_literal_intersection(tag1, pos1, tag2, pos2)
+        [{tag, fields, negs1 ++ negs2} | acc]
+      catch
+        :empty_intersection -> acc
+      end
+    end)
+  end
+
+  # Intersects two map literals; throws if their intersection is empty.
+  defp map_literal_intersection(tag1, map1, tag2, map2) do
+    default1 = if tag1 == :open, do: @term_or_unset, else: @unset
+    default2 = if tag2 == :open, do: @term_or_unset, else: @unset
+
+    # if any intersection of values is empty, the whole intersection is empty
+    new_fields =
+      (for {key, value_type} <- map1 do
+         value_type2 = Map.get(map2, key, default2)
+         t = intersection(value_type, value_type2)
+         if empty?(t), do: throw(:empty_intersection), else: {key, t}
+       end ++
+         for {key, value_type} <- map2, not is_map_key(map1, key) do
+           t = intersection(default1, value_type)
+           if empty?(t), do: throw(:empty_intersection), else: {key, t}
+         end)
+      |> Map.new()
+
+    case {tag1, tag2} do
+      {:open, :open} -> {:open, new_fields}
+      _ -> {:closed, new_fields}
+    end
+  end
+
+  defp map_difference(dnf1, dnf2) do
+    case dnf2 do
+      [] ->
+        dnf1
+
+      [lit2 | dnf2] ->
+        Enum.flat_map(dnf1, fn lit1 -> map_single_difference(lit1, lit2) end)
+        |> map_difference(dnf2)
+    end
+  end
+
+  # Computes the difference between two maps union.
+  defp map_single_difference({tag1, fields1, negs1}, {tag2, fields2, negs2}) do
+    Enum.reduce(negs2, [{tag1, fields1, [{tag2, fields2} | negs1]}], fn
+      {tag2, fields2}, acc ->
+        try do
+          {tag, fields} = map_literal_intersection(tag1, fields1, tag2, fields2)
+          [{tag, fields, negs1} | acc]
+        catch
+          :empty_intersection -> acc
+        end
+    end)
+  end
+
+  # Emptiness checking for maps. Short-circuits if it finds a non-empty map literal in the union.
+  defp map_empty?(dnf) do
+    try do
+      for {tag, pos, negs} <- dnf do
+        map_empty?(tag, pos, negs)
+      end
+
+      true
+    catch
+      :not_empty -> false
+    end
+  end
+
+  defp map_empty?(_, _, []), do: throw(:not_empty)
+  defp map_empty?(_, _, [{:open, neg_fields} | _]) when neg_fields == %{}, do: true
+  defp map_empty?(:open, fs, [{:closed, _} | negs]), do: map_empty?(:open, fs, negs)
+
+  defp map_empty?(tag, fields, [{neg_tag, neg_fields} | negs]) do
+    try do
+      # keys that are present in the negative map, but not in the positive one
+      for {neg_key, neg_type} <- neg_fields, not is_map_key(fields, neg_key) do
+        cond do
+          # key is required, and the positive map is closed: empty intersection
+          tag == :closed and not is_optional?(neg_type) ->
+            throw(:no_intersection)
```

Comment:
I think if we are "looping" just to find if a condition is satisfied, we should consider [`Enum.any?/2`](https://hexdocs.pm/elixir/main/Enum.html#any?/2)  rather than a `for` with `throw` which is usually [discouraged](https://hexdocs.pm/elixir/main/try-catch-and-rescue.html#throws).
`Enum.any` will also bail at the first truthy return so should be exactly what you want.

Another benefit is that otherwise your second list will build a list which is wasteful since you don't care about the list.

Reply:
Will send a follow-up PR for these!

Reply:
I think there is a bug in this cond. The tag can be closed and the type is optional.